### DISCUSSION TITLE
Cloudless

### DISCUSSION THREAD

**Discussion by ChessSpider** [Tasshack/dreame-vacuum#323] [Q&A] [answered]
Hello everyone,

I'm currently looking into integrating my Dreame L10s Ultra with Home Assistant and am interested in operating it without relying on cloud services. I have a few questions and would greatly appreciate your insights:

* Is it possible to set up a direct connection between the Dreame L10s Ultra and Home Assistant using the dreame-vacuum integration, bypassing the need for cloud connectivity?
* After setting up, can I safely block internet access to the vacuum cleaner to ensure it only operates within my local network?

I'm curious to know if anyone has successfully achieved a similar setup or if I should consider alternative methods like using Valetudo. Your experiences and suggestions will be very helpful.

Thank you!

**Comment by Tasshack** [comment, marked as answer]
Controlling the device over local network only possible when the device installed to the Mi Home cloud because it disables the local api when it is installed to the Dreamehome cloud also it does not reenable the api without hard reseting. Declouding the models that are not available in the Mi Home app is not possible because of this.

All the requests to the vacuum are made over local network when device is installed to the Mi Home app (unless `Prefer Cloud` option has been selected). Cloud connection is only required for retrieving the map data so if you choose to block the device from the internet you will loose map feature with its automaticaly generated room and map entities. Valetudo is the only solution that provides map feature without cloud connection so far.

I don't think this integration would be usefull without the map feature for advanced device like a L10s but if you are ready to give up on the map feature for the sake of declouding the device, here are the things you need to do;
- Hard reset the device if you are currently using the Dreamehome app and install the device to the Mi Home app. (You will loose all settings, maps and cleaning statistics)
- Get your local api token using [Xiaomi Cloud Tokens Extractor](https://github.com/PiotrMachowski/Xiaomi-cloud-tokens-extractor)
- Select `Without map` option while adding the device to the Home Assistant
- Enter the token you have extracted and add the device to the Home Assistant
- Fast map the house and create a disabled schedule for accessing the room ids from vacuum attributes.
- Block device internet connection with `Reject` option not `Drop`
- Also please don't forget that map changes over time and it may become corrupt because of the reflective surfaces in the house so you have to constantly restore the map to a previously known state using the integrations offline map recovery feature which has been added to Beta version. (Check the Beta docs)

I hope this information helps anyone who wants to decloud their device.

**Reply by Khraut** [reply]
I think I found a possible workaround for Dreame robots. It seems like that the devices used via the Dreamehome App only need to have incoming traffic while you can block outgoing traffic and it still works. Can somebody replicate this behaviour?

**Reply by Tasshack** [reply]
If it works that would mean you are not blocking anything from cloud so what is the point of this.

**Reply by Khraut** [reply]
Wouldn't that mean that no data is leaving my network? At least that is what i was thinking. I can control the robot but can't acces the camera feed. I also don't see the realtime position anymore so I was assuming that it is a compromise.

**Reply by Tasshack** [reply]
Have you restarted your vacuum?

**Reply by Khraut** [reply]
Yes I did.

**Reply by Tasshack** [reply]
If you can control the vacuum from app that would mean robot is still connected to the cloud via a socket connection but map data is stored on a cloud storage which uses a different https api so I guess you have just blocked the map data upload.

**Reply by Khraut** [reply]
You are right, i forgot to block the other connections. Sorry for the disruption and misleading info!

**Reply by Tasshack** [reply]
Also please note that, there are no incoming connections from cloud to the vacuum. It's the vacuum that connects to the cloud not the other way around.

**Reply by Patrick1610** [reply]
Hi! Is it possible to add the Dreame L10S Gen 2 back into the Dream App after getting the token? Will the token ever change or is it fixed?

**Reply by Tasshack** [reply]
> Hi! Is it possible to add the Dreame L10S Gen 2 back into the Dream App after getting the token? Will the token ever change or is it fixed?

Device disables local api when it is installed to the Dreame cloud.

**Reply by Patrick1610** [reply]
Thanks, that's very unfortunate! I really like the Dreame app itself and all the functions, but I think I want more flexibility in the schedules so I am considering Home Assistant. I am wondering: will I lose any functionality in the Xiaomi app, or is it nearly identical to Dreame itself? I have a L10S Ultra Gen 2 if that matters!

**Reply by Tasshack** [reply]
I don't think that device can be added to the Mi Home app.

**Reply by lockheed** [reply]
It's been two years since the statement cloud is necessary for maps to work. Has anything changed in this respect? Is it makes a difference, I'm talking about Dreame L10s Ultra, Ultra Heat and Mova P10 Pro Ultra

**Reply by Tasshack** [reply]
I have no idea why you think it will magically work after a certain period of time has passed.
Dreame does not want their devices to work without cloud and that did not change over the past years.

**Reply by lockheed** [reply]
I have no idea why you think I thought it would magically work instead of developer effort or manufacturer's change of position. But thanks for clearing it up.

**Reply by Tasshack** [reply]
Declouding is not aim of the project so there is no developer efford towards that in here.

**Reply by IdiotTryingToFindWay** [reply]
@Tasshack  if i flash valetudo on my dreame, will it still work correctly with your integration? like is there different steps that i would need to go through when setting it up.

**Reply by Tasshack** [reply]
Hypfer would be very upset about that...
He deliberately blocked people and their work to prevent Valetudo compatibility of this project.

https://github.com/Tasshack/dreame-vacuum/pull/1093

**Reply by rugk** [reply]
For reference: The _Mi Home_ app is not the same as the [_Xiaomi app_ ](https://play.google.com/store/apps/details?id=com.xiaomi.smarthome) is it?
If so, where does one get it?

**Comment by rugk** [comment]
## overview of robots listed in Xiaomi Home

I tried to connect the X50 (r2532d) to the Xiaomi Home app, but the app does not find the device. Also when scanning the QR code it only suggests to add the device manually.

When manually adding you see these devices listed there for vacuum robots (for reference to find a potential compatible device?):
<details>
<summary>big screenshots</summary>

<img width="1014" height="7364" alt="Screenshot_20260807-120703" src="https://github.com/user-attachments/assets/326dea6a-be9e-4638-b734-bc78d612629e" />
<img width="1014" height="6312" alt="Screenshot_20260807-121202" src="https://github.com/user-attachments/assets/5a9e0fac-f9cd-4a75-bc64-5a09393336e9" />
<img width="1014" height="6312" alt="Screenshot_20260807-121225" src="https://github.com/user-attachments/assets/5db37f1f-2b8f-4182-9908-d95ab924fe3f" />
<img width="1080" height="2340" alt="Screenshot_20260807-121331" src="https://github.com/user-attachments/assets/f0a06f8e-a437-40e0-b0cd-a617b9379028" />
<img width="1080" height="2159" alt="IMG_20260807_121453" src="https://github.com/user-attachments/assets/04306da4-784c-4467-a72e-2a5f8c76bedc" />
<img width="1080" height="2340" alt="Screenshot_20260807-122351" src="https://github.com/user-attachments/assets/f6723703-ee09-455f-afaa-3deb3fc3ff8d" />


</details>

These seem to be all. I don't get why no X50 is in there and connecting does not work hmm.

Also note, pressing 7 seconds both buttons on the robot does make it open its wifi for network config, and you can even connect with the phone and it says "phone connected, please return to the app to finish", but the app does not react upon this.
Evne if you select a similar module. After all, I would have guessed, the pairing should mostly work the same. But it somehow does not proceed?

**Comment by rugk** [comment]
# X50 problem (dreame.vacuum.r2532d) 

To open a new thread to keep my stuff about the X50 Ultra Complete in one place. [As explained above](https://github.com/Tasshack/dreame-vacuum/discussions/323#discussioncomment-17932392), it does not actually seem to pair with the Xiaomi Home app, nor is it shown in the list of robots in that app. 

One thing that wonders me then though is, when you scan the QR code on the device and just go to the website, you land here:
https://app.dreame.tech/?model=dreame.vacuum.r2532d

(I removed some sensitive query parameters containing the serial number and one other thing.)

<img width="1080" height="2340" alt="Screenshot_20260807-123822" src="https://github.com/user-attachments/assets/2196103c-946c-4eb2-bd94-38d494a19f1d" />


As for Android, this does list the Mi Home app (the orange one)? Why does not it work then? Or is the Xiaomu Home app (I have v11.6.704, 
`com.xiaomi.smarthome`) actually something different than Mi Home?

**Reply by rugk** [reply]
Bachlink https://github.com/PiotrMachowski/Xiaomi-cloud-tokens-extractor/discussions/200

**Reply by rugk** [reply]
For anyone later finding and reading this:
> Dreame devices manufactured after 2022 cannot be added to the Miio Cloud (Xiaomi Home) environment. 

https://community.home-assistant.io/t/are-there-any-100-local-control-robot-vacuums/616362/15?u=rugk
